### PR TITLE
Use PEP 503 compatible extra url index to install PyTorch

### PR DESCRIPTION
**Pull request by JanuszL**
- adds an option to list all extra indices for all installed packages
  in setup_packages.py
- thanks to this change torch whl can be fully downloaded and cached
  for the later installation process (as direct links used before are not)

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds the usage of PEP 503 compatible extra url index to install PyTorch

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     adds an option to list all extra indices for all installed packages in setup_packages.py
     thanks to this change torch whl can be fully downloaded and cached for the later installation process (as direct links used before are not)
 - Affected modules and functionalities:
     setup_packages.py, test_template.sh, 
 - Key points relevant for the review:download_pip_packages.sh
     NA
 - Validation and testing:
     current tests apply
 - Documentation (including examples):
     NA

**JIRA TASK**: *[DALI-2161]*

### CODE REVIEW

File: qa/setup_packages.py [hunk 1]
```diff
@@ -411,14 +458,16 @@ def cal_num_of_configs(packages, cuda_version):
             ret *= pckg.get_num_of_version(cuda_version)
     return ret
 
-def for_all_pckg(packages, fun):
-    """Iterates over all packages, executes a fun returns all fun results as a list"""
+def for_all_pckg(packages, fun, add_additional_packages=True):
+    """Iterates over all packages, executes a fun. Returns all fun results as a list"""
```
Comment (jantonguirao) [inline qa/setup_packages.py:462]:
```suggestion
    """Iterates over all packages, executes a function. Returns all function results as a list"""
```
"All fun results" read weird :)

File: qa/setup_packages.py [hunk 2]
```diff
@@ -263,6 +273,40 @@ def max_cuda_version(self, cuda_version):
...
+            Extra url used as pep 503 compatible repository to obtain listed packages
+    """
+    def __init__(self, key, versions, name=None, extra_index=""):
+        super(CudaPackageExtraIndex, self).__init__(key, versions, name)
+        if not isinstance(versions, dict):
+            raise TypeError("versions argument should by dict type [cuda_version : list_of_versions")
+        self.extra_index = extra_index
+
+    def get_extra_index(self, cuda_version):
+        """Gets a extra url index for pip for given cuda version.
```
Comment (jantonguirao) [inline qa/setup_packages.py:300]:
```suggestion
        """Gets a extra url index for pip for a given cuda version.
```

File: qa/setup_packages.py [hunk 3]
```diff
@@ -263,6 +273,40 @@ def max_cuda_version(self, cuda_version):
...
+        `name`: str, optional, default = None
+            Name of this package used during installation. If empty it is the same as the key.
+            If it includes `{cuda_v}` it is replaced by the cuda_version when queried
+        `extra_index`: str, optional, default = ""
+            Extra url used as pep 503 compatible repository to obtain listed packages
+    """
+    def __init__(self, key, versions, name=None, extra_index=""):
+        super(CudaPackageExtraIndex, self).__init__(key, versions, name)
+        if not isinstance(versions, dict):
+            raise TypeError("versions argument should by dict type [cuda_version : list_of_versions")
```
Comment (jantonguirao) [inline qa/setup_packages.py:296]:
```suggestion
            raise TypeError("versions argument should be a dictionary {cuda_version_str : list_of_versions}")
```

File: qa/setup_packages.py [hunk 4]
```diff
@@ -263,6 +273,40 @@ def max_cuda_version(self, cuda_version):
                 max_cuda = ver
         return max_cuda
 
+
+class CudaPackageExtraIndex(CudaPackage):
+    """Class describing a cuda package with a key/name and a dictionary where the key
+        is a cuda version and value is the list of versions supported.
```
Comment (jantonguirao) [inline qa/setup_packages.py:279]:
```suggestion
        is a cuda version and value is the list of versions supported.
```

**Review by jantonguirao — APPROVED** [review]

File: qa/setup_packages.py (hunk 3 above)
Comment (JanuszL) [reply qa/setup_packages.py]: Done

File: qa/setup_packages.py (hunk 2 above)
Comment (JanuszL) [reply qa/setup_packages.py]: Done

File: qa/setup_packages.py (hunk 1 above)
Comment (JanuszL) [reply qa/setup_packages.py]: Done

File: qa/setup_packages.py (hunk 4 above)
Comment (JanuszL) [reply qa/setup_packages.py:279]: ?

Comment (jantonguirao) [reply qa/setup_packages.py:279]: nevermind, I edited it back (thought the suggestion will be removed)

**Review by jantonguirao — APPROVED** [review]

File: qa/setup_packages.py (hunk 4 above)
Comment (JanuszL) [reply qa/setup_packages.py:279]: Ok

**Review by szalpal — APPROVED** [review]

**Pull request merged**